### PR TITLE
Synthetic Data Component Added (Backend Commit) / Classification and Regression

### PR DESCRIPTION
**Pull request by CodeByKarthik**
Hi Guys, 

This pull request introduces new scripts for generating synthetic data for classification and regression tasks.

Key changes mentioned below: 

- Added synthetic_data.py

Implements the CreateSyntheticData class to handle synthetic data generation

- synthetic_data_opts.py

Defines the SyntheticDataOptions dataclass to manage options for synthetic data creation.

- assertion.py

data_loader_asserts.py:

Added assertions to validate synthetic data options. 

- data.py

Modified data.py for integrating the synthetic data component. 

Kindly review the code and do the needful. 

Thanks and regards, 
Karthik

### CODE REVIEW

File: biofefi/machine_learning/data.py
Comment (dcl10) [inline biofefi/machine_learning/data.py:1]: I don't think that the synthetic data should be generated in `DataBuilder` class. It is unhelpfully named, I'll admit. It's purpose is to load in a data file and then normalise and bootstrap the data. What I would do is use the class you made above to generate the synthetic data. You can then save that data and then load that synthetic data back in with the `DataBuilder` class like any user provided dataset

**Review by dcl10 — CHANGES_REQUESTED** [review]

File: biofefi/options/synthetic_data_opts.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+@dataclass
+class SyntheticDataOptions:
+
+    # --- Problem Type Options ---
+    problem_type: str = (
+        ProblemTypes.Classification
+        if ProblemTypes.Classification
+        else ProblemTypes.Regression
+    )
```
[the remark this answers is not in the document]
Comment (Karthi-DStech) [reply biofefi/options/synthetic_data_opts.py:15]: Ok Dan

File: biofefi/options/synthetic_data_opts.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    random_state: int = 46
+
+    # --- Classification Options ---
+    n_features: int = 20
+    n_informative: int = 2
+    n_redudant: int = 2
+    n_repeated: int = 0
+    n_classes: int = 2
+    n_clusters_per_class: int = 2
+    data_weights: Optional[List[float]] = None
```
[the remark this answers is not in the document]
Comment (Karthi-DStech) [reply biofefi/options/synthetic_data_opts.py:30]: Ok

File: biofefi/options/synthetic_data_opts.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    flip_y: float = 0.01
+    class_sep: float = 1.0
+    data_scale: Optional[float] = None
+
+    # --- Regression Options ---
+    reg_n_informative: int = 10
+    reg_n_targets: int = 1
+    reg_bias: float = 0.0
+    reg_noise: float = 0.0
+    reg_coef: Optional[bool] = False
```
[the remark this answers is not in the document]
Comment (Karthi-DStech) [reply biofefi/options/synthetic_data_opts.py:40]: Ok

File: biofefi/machine_learning/data.py
Comment (Karthi-DStech) [reply biofefi/machine_learning/data.py:1]: Ok, Dan. I'll create an option for creating synthetic data and saving it into a CSV. This should allow users to load the data through a GUI.

File: biofefi/components/synthetic_data.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+            # Save the synthetic data to a csv file
+            synthetic_data.to_csv(self._output_file_name, index=False)
+            self._logger.info(f"Synthetic data saved to {self.output_file}")
+
+        except Exception as e:
+            raise InitialisationError(f"An Error initialising synthetic data: {e}")
+
+
+class InitialisationError(Exception):
```
Comment (dcl10) [inline biofefi/components/synthetic_data.py:143]:
```suggestion
class SyntheticDataInitialisationError(Exception):
```
`InitialisationError` sounds too generic

File: biofefi/components/synthetic_data.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                return pd.DataFrame(X), pd.Series(y, name="target")
+            except Exception as e:
+                raise ValueError(f"Error creating synthetic data for regression: {e}")
+
+    def data_initialisation(self) -> None:
+        """
+        Initialises the synthetic data creation.
+
+        Raises:
+            - InitialisationError: If an error occurs during
```
Comment (dcl10) [inline biofefi/components/synthetic_data.py:113]:
```suggestion
            - SyntheticDataInitialisationError: If an error occurs during
```

File: biofefi/components/synthetic_data.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            # Perform assertions on the generated data
+            data_checker = DataLoaderChecker(self.X, self.y)
+            data_checker.perform_data_checks()
+
+            # Save the synthetic data to a csv file
+            synthetic_data.to_csv(self._output_file_name, index=False)
+            self._logger.info(f"Synthetic data saved to {self.output_file}")
+
+        except Exception as e:
+            raise InitialisationError(f"An Error initialising synthetic data: {e}")
```
Comment (dcl10) [inline biofefi/components/synthetic_data.py:140]:
```suggestion
            raise SyntheticDataInitialisationError(f"An Error initialising synthetic data: {e}")
```

File: biofefi/components/synthetic_data.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
+import pandas as pd
+from sklearn.datasets import make_classification, make_regression
+
+from biofefi.options.enums import ProblemTypes
+from biofefi.options.synthetic_data_opts import SyntheticDataOptions
+from biofefi.utils.assertion import DataLoaderChecker
+
+
+class CreateSyntheticData:
```
Comment (dcl10) [inline biofefi/components/synthetic_data.py:9]:
```suggestion
class SyntheticDataBuilder:
```
Class names should be nouns

File: biofefi/machine_learning/data.py [hunk 8]
```diff
@@ -34,20 +35,49 @@ def __init__(
...
         self._n_bootstraps = n_bootstraps
+        self._create_synthetic_data = False
+
+    class DataBuilderError(Exception):
+        """
+        Custom exception raised when data loading or
+        generation fails in DataBuilder.
+        """
+
+        pass
```
Comment (dcl10) [inline biofefi/machine_learning/data.py:46]: Can we move this outside the `DataBuilder` class please?

**Review by dcl10 — CHANGES_REQUESTED** [review]

File: biofefi/components/synthetic_data.py (hunk 4 above)
Comment (Karthi-DStech) [reply biofefi/components/synthetic_data.py:143]: Ok Dan. I forget in the last commit I guess. My bad

File: biofefi/components/synthetic_data.py (hunk 5 above)
Comment (Karthi-DStech) [reply biofefi/components/synthetic_data.py:113]: Ok

File: biofefi/components/synthetic_data.py (hunk 6 above)
Comment (Karthi-DStech) [reply biofefi/components/synthetic_data.py:140]: Ok

File: biofefi/components/synthetic_data.py (hunk 7 above)
Comment (Karthi-DStech) [reply biofefi/components/synthetic_data.py:9]: Ok Dan

File: biofefi/machine_learning/data.py (hunk 8 above)
Comment (Karthi-DStech) [reply biofefi/machine_learning/data.py:46]: Ok Dan.

**Review by dcl10 — DISMISSED** [review]

**Review by dcl10 — CHANGES_REQUESTED** [review]
Just those last two unnecessary problem type checks

File: biofefi/components/synthetic_data.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+            - Tuple: The synthetic data and labels.
+
+        Raises:
+            - ValueError: If an error occurs during the data creation.
+        """
+
+        if self._problem_type == ProblemTypes.Classification:
+            self._logger.info(
+                "Requested to create synthetic data for Classification..."
+            )
```
[the remark this answers is not in the document]
Comment (Karthi-DStech) [reply biofefi/components/synthetic_data.py:54]: Ok Dan.

**Comment by Karthi-DStech** [comment]
> Just those last two unnecessary problem type checks

Ok Dan.

**Review by dcl10 — APPROVED** [review]

**Comment by Karthi-DStech** [comment]
Ok Dan. I'll merge it on the main. 

May I know what would be the next step? Is it working on the front-end part of the code for Synthetic data and Bayesian Regularised Neural Network yeah or have you got any other tasks?

**Pull request merged**